update anova type II and III (#391)
openpharma/mmrm: vignettes/hypothesis_testing.Rmd

package: mmrm
bibliography: '`r system.file("REFERENCES.bib", package = "mmrm")`'
csl: '`r system.file("jss.csl", package = "mmrm")`'
output:
  rmarkdown::html_vignette:
          toc: true
vignette: |
  %\VignetteIndexEntry{Details of Hypothesis Testing}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  chunk_output_type: console
  markdown:
    wrap: 72
---

# Introduction to Type I/II/III Hypothesis Testing

Type I/II/III hypothesis testing originates from analysis of variance, however, 
@goodnight1980tests suggest viewing this issue as hypothesis testing of fixed effect
in linear models, which avoid the "usual assumptions" and defined three types of estimable
functions.
Given that type I estimable functions is dependent on the order of covariate we specify,
we focus on the type II and III testing here.

From the `mmrm` model perspective, obtaining estimable functions, is equivalent to obtaining the
contrasts.

For more details, see @goodnight1980tests.

# Contained Effect

Before we discuss the testing, it is important that we understand the concept of "contained effect".
The definition of "contained effect" is:

> For effect "E2", it is said to contain "E1" if

Reviewer: Can we rather use $E_2$ etc. that is more mathematical notation?

> 1. they both involve the same continuous variable (and no more other continuous variable is involved)

Reviewer: ```suggestion
1. they both involve the same continuous covariates
```
I guess we don't need the parenthesis? Or we need to reword to make it clearer?

> 1. "E2" have equal or more "categorical variable" than "E1"

Reviewer: ```suggestion
1. "E2" have equal or more categorical covariates than "E1"
```

> For example, for the following formula

Reviewer: ```suggestion
For example, for the following model formula
```

$Y = A + B + A*B$

we have

> 1. If $A$ and $B$ are both categorical, then $F_{A*B}$ contains $F_{A}$ and $F_{B}$.

Reviewer: please first define what $F_{X}$ means, otherwise this is not clear

1. If $A$ and $B$ are both numerical, then $F_{A*B}$ do not contain $F_{A}$ or $F_{B}$.
1. If $A$ is categorical and $B$ is numerical, then $F_{A*B}$ contains $F_{B}$ but not $F_{A}$.

# Type II Hypothesis Testing

For type II hypothesis testing, the contrasts can be obtained through the following steps:

Create a contrast matrix $L$, with rows equal to the number of parameters associated with the effect,
columns equal to the number of all parameters.

1. All columns of $L$ associated with effect not containing $F_1$ (except $F_1$) are set to 0.
1. The submatrix of $L$ associated with $F_1$ is set to $(X_1^\top M X_1)^{-}X_1^\top M X_1$
1. Each of the remaining submatrices of L associated with an effect $F_2$ that contains $F_1$ is $(X_1^\top M X_1)^{-}X_1^\top M X_2$

where $X_0$ stands for columns of $X$ whose associated effect do not contain $F_1$,
$X_1$ stands for columns of $X$ associated with $F_1$,
$X_2$ stands for columns of $X$ whose associated effect contains $F_1$,
$M = I - X_0(X_0^\top X_0)^{-}X_0^\top$,

> and $Z^{-}$ stands for the g2 inverse of $Z$.

Reviewer: ```suggestion
and $Z^{-}$ stands for the g2 inverse of $Z$.
```
please include a link to a g2 inverse definition (I don't think this is commonly known)

> Note: In `mmrm` package, we do not allow singularity in general, so the g2 inverse is the inverse.

Reviewer: ```suggestion
Note: Here we do not allow for singularity in general, so the g2 inverse is just the usual inverse.
```

> Thus we can use identity matrix in step 2.

Reviewer: ```suggestion
Thus we can use an identity matrix as the submatrix in step 2.
```

> 

Reviewer: can you add an example here? I think that would be helpful

# Type III Hypothesis Testing

For type III hypothesis testing, we assume that the hypothesis tested should be the same for all
designs with the same general form of estimable functions.

> It can be defined with the following step:

Reviewer: ```suggestion
It can be defined with the following steps:
```

Create a contrast matrix $L$, with rows equal to the number of parameters associated with the effect,
columns equal to the number of all parameters.

1. All columns of $L$ associated with effect not containing $F_1$ (except $F_1$) are set to 0.
1. The submatrix of $L$ associated with $F_1$ is set to identity matrix.
1. For each of the remaining submatrices of L associated with an effect $F_2$ that contains $F_1$, equate the
values.

Here equate the values means that, given that the coefficients in step 2 are all 1 (diagonal), in step 3
this 1 is divided by the total number of new levels.

> 

Reviewer: also here an example will be great

# Hypothesis Testing in SAS

In `PROC MIXED` we can specify `HTYPE=1,2,3` to enable these hypothesis testings, and we can use option `e1 e2 e3` to
view the estimable functions.

## Special Considerations

For `PROC MIXED`, it is important that the categorical values have the correct order, especially
for structured covariance (otherwise the correlation between visits can be messed up).
We usually use `CLASS variable(REF=)` to define the reference level.
However, SAS will put the reference level in the last, so if the visit variable is included in fixed effect,
there can be issues with the comparison between SAS and R, as in R we still use the first level as our reference.

When we conduct the hypothesis testing, we will also face this issue.

> The result will be slightly different because of different reference groups, however, if correct contrast is used

Reviewer: but is it possible always to match results by custom options for reference levels? or not?
would be good to make that clear

Author: I think it is already clear that `if correct contrast is used we will get closer result`. (I used closer because anyway we will not have identical results)

However, we do not include the algorithm to create the correct reference.

Reviewer: Hm it is not clear to me yet :-) Is the way we show the code currently getting us "closest" to SAS results? or is there another way to get even closer?
If there is an algorithm to create the correct reference/contrast then let's include it

Author: well this is not "correct reference" but user's choices. e.g. for RACE we really don't need to say "which reference is correct, Aisan or White"? they essentially try to answer the same question, however the result can be a little different. For type II it will be a little complicated to obtain the contrast for the with different reference level, for type III it is easier, e.g. if there are three levels, using lvl1 as reference, we are testing `lvl2 - lvl1` and `lvl3 - lvl1` are both zero, and the contrast should be
```
1 0
0 1
```

however if we are using `lvl3` as reference, then we should test `lvl1 - lvl3`, `lvl2 - lvl3`. However in R we have `lvl2-lvl1`, `lvl3 - lvl1`. With a linear combination of these we can use

```
0 1
-1 -1
```
So here I use "correct contrast" to say that, e.g. in SAS we are testing `lvl1 - lvl3`, `lvl2 - lvl3`, in R we shoud not use identity matrix if we only have `lvl2 - lvl1` and `lvl3 - lvl1` and identity matrix, but use the previous form of matrix to correctly test `lvl1 - lvl3`, `lvl2 - lvl3`. It is not related to "correct reference", but "correct contrast" to reflect the same testing.

However a better solution is to use a different model to create the same design matrix.

In summary, I think we should not provide such functionality to update the contrast afterwards

Reviewer: ok makes sense, can you add this example to the vignette? I think that helps the user

Author: updated

we should have closer result.

### Why Model Covariate Order Changes My Testing in SAS?

For type II/III testing, it is true that the order of covariate should not affect the result.
However, if there is an interaction term, the reference level can be changed.

> With different reference level the result will be slightly different.

Reviewer: ```suggestion
With different reference levels the result will be slightly different.
```

### Why `mmrm` Gives More Covariates Than SAS?

For `PROC MIXED`, if a model is specified as follows

```
MODEL FEV1 = ARMCD SEX ARMCD * SEX ARMCD * FEV1_BL
```

Then it is equivalent to the following `mmrm` model

```

> FEV1 ~ ARMCD * SEX + ARMCD * FEV1_BL - FEV1_BL

Reviewer: ```suggestion
FEV1 ~ ARMCD * SEX + ARMCD:FEV1_BL
```
would be simpler

Author: this is on purpose to use `ARMCD*FEV1_BL` to better illustrate the difference: in SAS using `ARMCD*FEV1_BL` will only add the interaction term, but not the two main effect.

Reviewer: can we then add a comment that in R you can do the "ARMCD:FEV1_BL" instead?

Author: added

```

Because SAS will not include the covariate `FEV1_BL` by default, unless manually added.
However in R we will add `FEV1_BL` by default.

> Please note that in this example we excludes the covariance structure part.

Reviewer: ```suggestion
Please note that in this example we exclude the covariance structure part.
```

1. Each of the remaining submatrices of L associated with an effect $E_2$ that contains $E_1$ is $(X_1^\top M X_1)^{-}X_1^\top M X_2$

where $X_0$ stands for columns of $X$ whose associated effect do not contain $E_1$,
$X_1$ stands for columns of $X$ associated with $E_1$,
$X_2$ stands for columns of $X$ whose associated effect contains $E_1$,
$M = I - X_0(X_0^\top X_0)^{-}X_0^\top$,
and $Z^{-}$ stands for the [g2 inverse](https://blogs.sas.com/content/iml/2018/11/21/generalized-inverses-for-matrices.html) of $Z$.

Note: Here we do not allow for singularity in general, so the g2 inverse is just the usual inverse.
Thus we can use an identity matrix as the submatrix in step 2.

Using `fev_data` to create our example, we have

```{r}
library(mmrm)
fit <- mmrm(FEV1 ~ ARMCD + RACE + ARMCD * RACE + ar1(AVISIT | USUBJID), data = fev_data)
```

For this given example, we would like to test the effect of `RACE`, $E_{RACE}$.
We initialize the contrast matrix with

$$
\begin{matrix} 
\mu & ARMCD & RACE & RACE & ARMCD:RACE & ARMCD:RACE\\
0 & 0 & 0 & 0 & 0 & 0\\ 
0 & 0 & 0 & 0 & 0 & 0\\ 
\end{matrix}
$$

> Please note that this is $2\times 6$ matrix, the rows is the number of coefficients that is associated with `RACE`,

Reviewer: ```suggestion
Please note that this is a $2\times 6$ matrix, the rows is the number of coefficients that is associated with `RACE`,
```

> and the columns is the total number of coefficients `r length(coef(fit))`

Reviewer: ```suggestion
and the columns is the total number of coefficients `r length(coef(fit))`.
```

> Then following step 2, we filled in identity matrix in the corresponding submatrix

Reviewer: ```suggestion
Then following step 2, we filled in the identity matrix in the corresponding submatrix
```

0 & 0 & 0 & 1 & -0.04372702 & 0.58570969\\ 
\end{matrix}
$$

# Type III Hypothesis Testing

For type III hypothesis testing, we assume that the hypothesis tested should be the same for all
designs with the same general form of estimable functions.
It can be defined with the following steps:

Create a contrast matrix $L$, with rows equal to the number of parameters associated with the effect,
columns equal to the number of all parameters.

1. All columns of $L$ associated with effect not containing $E_1$ (except $E_1$) are set to 0.
1. The submatrix of $L$ associated with $E_1$ is set to identity matrix.
1. For each of the remaining submatrices of L associated with an effect $E_2$ that contains $E_1$, equate the
values.

Here equate the values means that, given that the coefficients in step 2 are all 1 (diagonal), in step 3
this 1 is divided by the total number of new levels.

Using the same example for type II testing, we also would like to test the effect of `RACE`, $E_{RACE}$.
We initialize the contrast matrix with

$$
\begin{matrix} 
\mu & ARMCD & RACE & RACE & ARMCD:RACE & ARMCD:RACE\\
0 & 0 & 0 & 0 & 0 & 0\\ 
0 & 0 & 0 & 0 & 0 & 0\\ 
\end{matrix}
$$

> Then following step 2, we filled in identity matrix in the corresponding submatrix

Reviewer: ```suggestion
Then following step 2, we filled in the identity matrix in the corresponding submatrix
```

The following example illustrate this.

#### Example of Reference Levels

Assume in SAS we have the following model

```
PROC MIXED DATA = fev cl method=reml;
  CLASS AVISIT(ref = 'VIS4') ARMCD(ref = 'PBO') USUBJID;
  MODEL FEV1 = ARMCD AVISIT ARMCD*AVISIT / ddfm=satterthwaite htype=3;
  REPEATED AVISIT / subject=USUBJID type=ar(1);
RUN;
```

And in R we have the following model
```{r eval=FALSE}
fit <- mmrm(FEV1 ~ ARMCD + AVISIT + ARMCD * AVISIT + ar1(AVISIT | USUBJID), data = fev_data)
Anova(fit)
```

Note that for `AVISIT` there are 4 levels, `VIS1`, `VIS2`, `VIS3` and `VIS4`.
In SAS `VIS4` is the reference, while in R `VIS1` is the reference.

They both use the following matrix to test the effect of `AVISIT`
(ignoring other part of contrasts that is not associated with `AVISIT`)

$$
\begin{matrix} 
1 & 0 & 0\\ 
0 & 1 & 0\\
0 & 0 & 1\\
\end{matrix}
$$

But given that the reference level is different, we are testing different things.
In SAS we are testing `VIS1 - VIS4`, `VIS2 - VIS4` and `VIS3 - VIS4`,
while in R we are testing `VIS2 - VIS1`, `VIS3 - VIS1` and `VIS4 - VIS1`.

To correctly test `VIS1 - VIS4`, `VIS2 - VIS4` and `VIS3 - VIS4` in R, we need to update the contrast

> to the following matrix(ignoring other part) to get closer result.

Reviewer: ```suggestion
to the following matrix (ignoring the other part) to get a closer result.
```
